File: ChessLab/Maia/MaiaEncoder.swift
import ChessKit
import Foundation

/// Encodeur d'entrée de Maia-3 : transforme les dernières positions d'une
/// partie en tenseur `64 × 97` (une ligne par case).
///
/// Réplique exacte de `tokenize_board()` et `get_historical_tokens()` du dépôt
/// `CSSLab/maia3`, prouvée bit à bit par `MaiaFixtureTests` :
///
///  - **Une position = 12 plans par case** : pion, cavalier, fou, tour, dame,
///    roi du camp AU TRAIT (plans 0-5), puis les mêmes pour l'adversaire
///    (plans 6-11). Pas de droits de roque, pas d'en passant : le modèle les
///    infère de l'historique.
///  - **Le plateau est retourné quand les Noirs jouent** (rangée 9 − r,
///    couleurs échangées), position par position : dans l'historique, les
///    orientations alternent donc d'un demi-coup à l'autre. C'est ce que fait
///    le code de référence, pas une simplification.
///  - **Historique de 8 positions**, de la plus ancienne (colonnes 0-11) à la
///    courante (colonnes 84-95) ; s'il en manque, la plus ancienne connue est
///    répétée en tête. La 97e colonne (temps de réflexion) reste à zéro à
///    l'inférence.
enum MaiaEncoder {
    static let historyLength = 8
    static let planesPerPosition = 12
    static let featuresPerSquare = historyLength * planesPerPosition + 1  // 97
    static let squareCount = 64

    /// Numéro de plan d'une pièce, dans le repère du camp au trait.
    static func plane(of piece: Piece, sideToMove: Piece.Color) -> Int {
        let kindOffset: Int = switch piece.kind {
        case .pawn: 0
        case .knight: 1
        case .bishop: 2
        case .rook: 3
        case .queen: 4
        case .king: 5
        }
        return kindOffset + (piece.color == sideToMove ? 0 : 6)
    }

    /// Les 64 × 12 bits d'UNE position, case par case (a1 → h8 après
    /// renversement éventuel), sous forme d'un masque de 12 bits par case.
    static func planes(of position: Position) -> [UInt16] {
        var squares = [UInt16](repeating: 0, count: squareCount)
        let mirror = position.sideToMove == .black
        for piece in position.pieces {
            let square = mirror ? MaiaMoveTable.mirrored(piece.square) : piece.square
            squares[MaiaMoveTable.squareIndex(square)] |= UInt16(1) << UInt16(plane(of: piece, sideToMove: position.sideToMove))
        }
        return squares
    }

    /// Tenseur d'entrée `64 × 97`, ligne par ligne, prêt pour Core ML.
    ///
    /// - parameter history: les positions de la partie, de la plus ancienne à
    ///   la COURANTE (dernier élément = position à jouer). Seules les huit
    ///   dernières comptent ; une liste vide rend un tenseur nul.
    static func tokens(history: [Position]) -> [Float] {
        var tensor = [Float](repeating: 0, count: squareCount * featuresPerSquare)
        guard !history.isEmpty else { return tensor }
        let recent = Array(history.suffix(historyLength))
        let padding = historyLength - recent.count
        let planesByPosition = recent.map(planes(of:))

        for slot in 0..<historyLength {
            // Les fentes manquantes en tête répètent la plus ancienne connue.
            let source = slot < padding ? planesByPosition[0] : planesByPosition[slot - padding]
            let columnBase = slot * planesPerPosition
            for square in 0..<squareCount {
                let mask = source[square]
                guard mask != 0 else { continue }
                let rowBase = square * featuresPerSquare + columnBase
                for plane in 0..<planesPerPosition where mask & (UInt16(1) << UInt16(plane)) != 0 {
                    tensor[rowBase + plane] = 1
                }
            }
        }
        return tensor
    }

    /// Les 96 bits d'historique d'une case, en hexadécimal (24 caractères,
    /// colonne 0 en poids fort) — le format des fixtures de référence.
    static func hexRow(_ tensor: [Float], square: Int) -> String {
        var hex = ""
        let base = square * featuresPerSquare
        for nibble in 0..<24 {
            var value = 0
            for bit in 0..<4 {
                value = value << 1 | (tensor[base + nibble * 4 + bit] > 0.5 ? 1 : 0)
            }
            hex += String(value, radix: 16)
        }
        return hex
    }
}

File: android/maia/src/main/java/com/chesslab/maia/MaiaEncoder.kt
package com.chesslab.maia

import chesskit.Piece
import chesskit.Position

/**
 * Traduction Kotlin de `MaiaEncoder.swift`.
 *
 * Transforme les dernières positions d'une partie en tenseur `64 × 97`, une
 * ligne par case. Réplique de `tokenize_board()` et `get_historical_tokens()`
 * du dépôt `CSSLab/maia3` :
 *
 *  - **Une position = 12 plans par case** : pion, cavalier, fou, tour, dame,
 *    roi du camp AU TRAIT (plans 0-5), puis les mêmes pour l'adversaire
 *    (6-11). Ni roque ni prise en passant : le modèle les infère de
 *    l'historique.
 *  - **Le plateau est retourné quand les Noirs jouent**, position par
 *    position : dans l'historique, les orientations alternent donc d'un
 *    demi-coup à l'autre. C'est le code de référence, pas une simplification.
 *  - **Historique de 8 positions**, de la plus ancienne aux colonnes 0-11 à la
 *    courante aux colonnes 84-95 ; s'il en manque, la plus ancienne connue est
 *    répétée en tête. La 97e colonne (temps de réflexion) reste à zéro.
 */
object MaiaEncoder {
    const val HISTORY_LENGTH = 8
    const val PLANES_PER_POSITION = 12
    const val FEATURES_PER_SQUARE = HISTORY_LENGTH * PLANES_PER_POSITION + 1   // 97
    const val SQUARE_COUNT = 64

    /** Le numéro de plan d'une pièce, dans le repère du camp au trait. */
    fun plane(piece: Piece, sideToMove: Piece.Color): Int {
        val kindOffset = when (piece.kind) {
            Piece.Kind.pawn -> 0
            Piece.Kind.knight -> 1
            Piece.Kind.bishop -> 2
            Piece.Kind.rook -> 3
            Piece.Kind.queen -> 4
            Piece.Kind.king -> 5
        }
        return kindOffset + if (piece.color == sideToMove) 0 else 6
    }

    /** Les 64 × 12 bits d'UNE position, un masque de 12 bits par case. */
    fun planes(position: Position): IntArray {
        val squares = IntArray(SQUARE_COUNT)
        val mirror = position.sideToMove == Piece.Color.black
        for (piece in position.pieces) {
            val square = if (mirror) MaiaMoveTable.mirrored(piece.square) else piece.square
            squares[MaiaMoveTable.squareIndex(square)] =
                squares[MaiaMoveTable.squareIndex(square)] or (1 shl plane(piece, position.sideToMove))
        }
        return squares
    }

    /**
     * Le tenseur d'entrée `64 × 97`, ligne par ligne.
     *
     * [history] va de la plus ancienne position à la COURANTE (dernier
     * élément = position à jouer). Seules les huit dernières comptent ; une
     * liste vide rend un tenseur nul.
     */
    fun tokens(history: List<Position>): FloatArray {
        val tensor = FloatArray(SQUARE_COUNT * FEATURES_PER_SQUARE)
        if (history.isEmpty()) return tensor

        val recent = history.takeLast(HISTORY_LENGTH)
        val padding = HISTORY_LENGTH - recent.size
        val planesByPosition = recent.map { planes(it) }

        for (slot in 0 until HISTORY_LENGTH) {
            // les fentes manquantes en tête répètent la plus ancienne connue
            val source = if (slot < padding) planesByPosition[0] else planesByPosition[slot - padding]
            val columnBase = slot * PLANES_PER_POSITION
            for (square in 0 until SQUARE_COUNT) {
                val mask = source[square]
                if (mask == 0) continue
                val rowBase = square * FEATURES_PER_SQUARE + columnBase
                for (plane in 0 until PLANES_PER_POSITION) {
                    if (mask and (1 shl plane) != 0) tensor[rowBase + plane] = 1f
                }
            }
        }
        return tensor
    }
}
